R/rdeephaven/lib/cpp-dependencies/build-dependencies.sh

```diff
@@ -0,0 +1,390 @@
+#!/bin/bash
+
+#
+# Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
+#
+
+#
+# Tested on Ubuntu 20.04
+#
+
+# Fail on first error; echo each command before executing.
+set -eux
+
+clean="no"
+if [ "$#" -ge 1 ]; then
+    if [ "$1" == "--clean" ]; then
+        clean="yes"
+        shift
+    fi
+fi
+
+step=all
+if [ "$#" -gt 0 ]; then
+    if [ "$#" -gt 1 ]; then
+        echo "Usage: $0 [checkout|protobuf|re2|gflags|absl|flatbuffers|cares|zlib|grpc|arrow|immer|boost|env]"
+        exit 1
+    fi
+    step=$1
+fi
+
+# Add anything to PATH that should take precendence here.
+# export PATH=/l/cmake/3.21.2/bin:/l/gcc/11.2.0/bin:$PATH
+
+# Edit to reflect your compiler preferences, or comment out for using system versions.
+# export CC=/l/gcc/11.2.0/bin/gcc
+# export CXX=/l/gcc/11.2.0/bin/g++
+
+# Set to "Debug" or "Release"
+: ${BUILD_TYPE:=Debug}
+
+# Set to where you intend the sources for and installed depdenencies to live.
+: ${DHDEPS_HOME:=$(pwd)}
+
+if [ "$step" = "all" ]; then
+# If you want to rebuild only certain parts or skip phases change "yes" to "no" below.
+# Note this assumues you at least once built everything; otherwise dependencies among libraries
+# may fail.
+  : ${BUILD_PROTOBUF:=yes}
+  : ${BUILD_RE2:=yes}
+  : ${BUILD_GFLAGS:=yes}
+  : ${BUILD_ABSL:=yes}
+  : ${BUILD_FLATBUFFERS:=no}
+  : ${BUILD_CARES:=yes}
+  : ${BUILD_ZLIB:=yes}
+  : ${BUILD_GRPC:=yes}
+  : ${BUILD_ARROW:=yes}
+  : ${BUILD_IMMER:=yes}
+  : ${GENERATE_ENV:=yes}
+else
+  BUILD_PROTOBUF=no
+  BUILD_RE2=no
+  BUILD_GFLAGS=no
+  BUILD_ABSL=no
+  BUILD_FLATBUFFERS=no
+  BUILD_CARES=no
+  BUILD_ZLIB=no
+  BUILD_GRPC=no
+  BUILD_ARROW=no
+  BUILD_IMMER=no
+  GENERATE_ENV=no
+  case "$step" in
+      protobuf)
+          BUILD_PROTOBUF=yes
+          ;;
+      re2)
+          BUILD_RE2=yes
+          ;;
+      gflags)
+          BUILD_GFLAGS=yes
+          ;;
+      absl)
+          BUILD_ABSL=yes
+          ;;
+      flatbuffers)
+          BUILD_FLATBUFFERS=yes
+          ;;
+      cares)
+          BUILD_CARES=yes
+          ;;
+      zlib)
+          BUILD_ZLIB=yes
+          ;;
+      grpc)
+          BUILD_GRPC=yes
+        ;;
+      arrow)
+          BUILD_ARROW=yes
+          ;;
+      immer)
+          BUILD_IMMER=yes
+          ;;
+      env)
+          GENERATE_ENV=yes
+          ;;
+      *)
+          echo "$0: unrecognized option: $step" 1>&2
+          exit 1
+          ;;
+  esac
+fi
+
+# At the point of this writing, the latest immer release is pretty old.
+# We want something a lot more recent, but don't want to track head as is a moving
+# target and we can't guarantee things will continue to compile/be consistent.
+# So we select a particular SHA.
+: ${IMMER_SHA:=e5d79ed80ec74d511cc4f52fb68feeac66507f2c}
```

Comment:
How are we going to keep this version in sync?

Reply:
Currently, all of the dependency versions are sprinkled all through this file.  It may be cleaner to consolidate them all here.